src/openrct2/world/Map.cpp

```diff
@@ -1539,6 +1542,92 @@ void map_update_tiles()
     }
 }
 
+void map_calc_highest_tile_height(bool fullCheck)
+{
+    int32_t maxTileIndex = MAXIMUM_MAP_SIZE_TECHNICAL * MAXIMUM_MAP_SIZE_TECHNICAL;
+    int32_t tilesToCheck = maxTileIndex;
+
+    // If we're doing a full check of the map then zero out the loop
+    // position to start at tile zero, checking all tiles before returning.
+    // Otherwise check a limited number of tiles per each time the
+    // function is called.
+    if (fullCheck)
+    {
+        gMapHighestTileHeightLoopPosition = 0;
+    }
+    else
+    {
+        tilesToCheck /= 128;
+    }
+
+    for (int32_t i = 0; i < tilesToCheck; i++)
+    {
+        gMapHighestTileHeightLoopPosition++;
+        gMapHighestTileHeightLoopPosition %= maxTileIndex;
+
+        if (gMapHighestTileHeightLoopPosition == 0)
+        {
+            gMapHighestTileHeight = gMapHighestTileHeightPending;
+            gMapHighestTileHeightPending = 0;
+        }
+
+        int32_t x = (gMapHighestTileHeightLoopPosition % MAXIMUM_MAP_SIZE_TECHNICAL) * COORDS_XY_STEP;
+        int32_t y = (gMapHighestTileHeightLoopPosition / MAXIMUM_MAP_SIZE_TECHNICAL) * COORDS_XY_STEP;
+        auto tileCoords = CoordsXY(x, y);
+        auto tileElement = map_get_first_element_at(tileCoords);
+
+        // Loop through elements on the current tile and find the tallest
+        do
+        {
+            if (tileElement == nullptr)
+                break;
+
+            if (tileElement->GetType() == TILE_ELEMENT_TYPE_CORRUPT)
```

Comment:
I'm only ignoring corrupted map elements. I had a look at the defined constants for element types, and it seemed like a good idea to include everything else in the check. Is there anything else that should be excluded?

Reply:
This all makes sense.

Reply:
> Is there anything else that should be excluded?

Do free-floating balloons count as map elements? Because they might get pretty high when floating away, and the scroll limits probably shouldn't change just because of balloons.

Reply:
Lol, I'm glad you brought that up. I'll check for balloons and ducks and make sure they're not included.

Reply:
A balloon isn't a tile element as far as I can tell. It looks like it's a different type of entity, so it shouldn't affect this logic at all.

Reply:
Balloons are entities. Not relevant.